In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

I have a dataset regarding tickets submitted to a customer helpdesk for an outlet that deals in electronic and technical goods. The first thing I will do is import the CSV as a Pandas dataframe.

In [2]:
df = pd.read_csv('customer_support_tickets.csv')
df.head()

Ticket ID        Customer Name              Customer Email  Customer Age  \
0          1        Marisa Obrien  carrollallison@example.com            32   
1          2         Jessica Rios    clarkeashley@example.com            42   
2          3  Christopher Robbins   gonzalestracy@example.com            48   
3          4     Christina Dillon    bradleyolson@example.org            27   
4          5    Alexander Carroll     bradleymark@example.com            67   

  Customer Gender Product Purchased Date of Purchase      Ticket Type  \
0           Other        GoPro Hero       2021-03-22  Technical issue   
1          Female       LG Smart TV       2021-05-22  Technical issue   
2           Other          Dell XPS       2020-07-14  Technical issue   
3          Female  Microsoft Office       2020-11-13  Billing inquiry   
4          Female  Autodesk AutoCAD       2020-02-04  Billing inquiry   

             Ticket Subject  \
0             Product setup   
1  Peripheral compatibility   
2           Network problem   
3            Account access   
4                 Data loss   

                                  Ticket Description  \
0  I'm having an issue with the {product_purchase...   
1  I'm having an issue with the {product_purchase...   
2  I'm facing a problem with my {product_purchase...   
3  I'm having an issue with the {product_purchase...   
4  I'm having an issue with the {product_purchase...   

               Ticket Status                                     Resolution  \
0  Pending Customer Response                                            NaN   
1  Pending Customer Response                                            NaN   
2                     Closed   Case maybe show recently my computer follow.   
3                     Closed  Try capital clearly never color toward story.   
4                     Closed                    West decision evidence bit.   

  Ticket Priority Ticket Channel  First Response Time   Time to Resolution  \
0        Critical   Social media  2023-06-01 12:15:36                  NaN   
1        Critical           Chat  2023-06-01 16:45:38                  NaN   
2             Low   Social media  2023-06-01 11:14:38  2023-06-01 18:05:38   
3             Low   Social media  2023-06-01 07:29:40  2023-06-01 01:57:40   
4             Low          Email  2023-06-01 00:12:42  2023-06-01 19:53:42   

   Customer Satisfaction Rating  
0                           NaN  
1                           NaN  
2                           3.0  
3                           3.0  
4                           1.0

I want to remove some customer information to ensure privacy. I will start by removing the email addresses. We can see that a lot of them end in some variation of 'example', so it may be that these are not real people, or their email addresses have already been altered. It's good practice to remove this info anyway, so let's proceed.

In [3]:
df.drop('Customer Email', axis=1, inplace=True)
df.head()

Ticket ID        Customer Name  Customer Age Customer Gender  \
0          1        Marisa Obrien            32           Other   
1          2         Jessica Rios            42          Female   
2          3  Christopher Robbins            48           Other   
3          4     Christina Dillon            27          Female   
4          5    Alexander Carroll            67          Female   

  Product Purchased Date of Purchase      Ticket Type  \
0        GoPro Hero       2021-03-22  Technical issue   
1       LG Smart TV       2021-05-22  Technical issue   
2          Dell XPS       2020-07-14  Technical issue   
3  Microsoft Office       2020-11-13  Billing inquiry   
4  Autodesk AutoCAD       2020-02-04  Billing inquiry   

             Ticket Subject  \
0             Product setup   
1  Peripheral compatibility   
2           Network problem   
3            Account access   
4                 Data loss   

                                  Ticket Description  \
0  I'm having an issue with the {product_purchase...   
1  I'm having an issue with the {product_purchase...   
2  I'm facing a problem with my {product_purchase...   
3  I'm having an issue with the {product_purchase...   
4  I'm having an issue with the {product_purchase...   

               Ticket Status                                     Resolution  \
0  Pending Customer Response                                            NaN   
1  Pending Customer Response                                            NaN   
2                     Closed   Case maybe show recently my computer follow.   
3                     Closed  Try capital clearly never color toward story.   
4                     Closed                    West decision evidence bit.   

  Ticket Priority Ticket Channel  First Response Time   Time to Resolution  \
0        Critical   Social media  2023-06-01 12:15:36                  NaN   
1        Critical           Chat  2023-06-01 16:45:38                  NaN   
2             Low   Social media  2023-06-01 11:14:38  2023-06-01 18:05:38   
3             Low   Social media  2023-06-01 07:29:40  2023-06-01 01:57:40   
4             Low          Email  2023-06-01 00:12:42  2023-06-01 19:53:42   

   Customer Satisfaction Rating  
0                           NaN  
1                           NaN  
2                           3.0  
3                           3.0  
4                           1.0

I'll also set the Ticket ID as the index to make the dataframe easier to read.

In [4]:
df.set_index('Ticket ID', inplace=True)

In [5]:
df.head()

Customer Name  Customer Age Customer Gender  \
Ticket ID                                                      
1                Marisa Obrien            32           Other   
2                 Jessica Rios            42          Female   
3          Christopher Robbins            48           Other   
4             Christina Dillon            27          Female   
5            Alexander Carroll            67          Female   

          Product Purchased Date of Purchase      Ticket Type  \
Ticket ID                                                       
1                GoPro Hero       2021-03-22  Technical issue   
2               LG Smart TV       2021-05-22  Technical issue   
3                  Dell XPS       2020-07-14  Technical issue   
4          Microsoft Office       2020-11-13  Billing inquiry   
5          Autodesk AutoCAD       2020-02-04  Billing inquiry   

                     Ticket Subject  \
Ticket ID                             
1                     Product setup   
2          Peripheral compatibility   
3                   Network problem   
4                    Account access   
5                         Data loss   

                                          Ticket Description  \
Ticket ID                                                      
1          I'm having an issue with the {product_purchase...   
2          I'm having an issue with the {product_purchase...   
3          I'm facing a problem with my {product_purchase...   
4          I'm having an issue with the {product_purchase...   
5          I'm having an issue with the {product_purchase...   

                       Ticket Status  \
Ticket ID                              
1          Pending Customer Response   
2          Pending Customer Response   
3                             Closed   
4                             Closed   
5                             Closed   

                                              Resolution Ticket Priority  \
Ticket ID                                                                  
1                                                    NaN        Critical   
2                                                    NaN        Critical   
3           Case maybe show recently my computer follow.             Low   
4          Try capital clearly never color toward story.             Low   
5                            West decision evidence bit.             Low   

          Ticket Channel  First Response Time   Time to Resolution  \
Ticket ID                                                            
1           Social media  2023-06-01 12:15:36                  NaN   
2                   Chat  2023-06-01 16:45:38                  NaN   
3           Social media  2023-06-01 11:14:38  2023-06-01 18:05:38   
4           Social media  2023-06-01 07:29:40  2023-06-01 01:57:40   
5                  Email  2023-06-01 00:12:42  2023-06-01 19:53:42   

           Customer Satisfaction Rating  
Ticket ID                                
1                                   NaN  
2                                   NaN  
3                                   3.0  
4                                   3.0  
5                                   1.0

Now I'll also drop the customer name as this also helps to make the data anonymous. We will not need the customer name for any of the data analysis.

In [6]:
df.drop('Customer Name', axis=1, inplace=True)
df.head()

Customer Age Customer Gender Product Purchased Date of Purchase  \
Ticket ID                                                                    
1                    32           Other        GoPro Hero       2021-03-22   
2                    42          Female       LG Smart TV       2021-05-22   
3                    48           Other          Dell XPS       2020-07-14   
4                    27          Female  Microsoft Office       2020-11-13   
5                    67          Female  Autodesk AutoCAD       2020-02-04   

               Ticket Type            Ticket Subject  \
Ticket ID                                              
1          Technical issue             Product setup   
2          Technical issue  Peripheral compatibility   
3          Technical issue           Network problem   
4          Billing inquiry            Account access   
5          Billing inquiry                 Data loss   

                                          Ticket Description  \
Ticket ID                                                      
1          I'm having an issue with the {product_purchase...   
2          I'm having an issue with the {product_purchase...   
3          I'm facing a problem with my {product_purchase...   
4          I'm having an issue with the {product_purchase...   
5          I'm having an issue with the {product_purchase...   

                       Ticket Status  \
Ticket ID                              
1          Pending Customer Response   
2          Pending Customer Response   
3                             Closed   
4                             Closed   
5                             Closed   

                                              Resolution Ticket Priority  \
Ticket ID                                                                  
1                                                    NaN        Critical   
2                                                    NaN        Critical   
3           Case maybe show recently my computer follow.             Low   
4          Try capital clearly never color toward story.             Low   
5                            West decision evidence bit.             Low   

          Ticket Channel  First Response Time   Time to Resolution  \
Ticket ID                                                            
1           Social media  2023-06-01 12:15:36                  NaN   
2                   Chat  2023-06-01 16:45:38                  NaN   
3           Social media  2023-06-01 11:14:38  2023-06-01 18:05:38   
4           Social media  2023-06-01 07:29:40  2023-06-01 01:57:40   
5                  Email  2023-06-01 00:12:42  2023-06-01 19:53:42   

           Customer Satisfaction Rating  
Ticket ID                                
1                                   NaN  
2                                   NaN  
3                                   3.0  
4                                   3.0  
5                                   1.0

In [7]:
df.describe()

Customer Age  Customer Satisfaction Rating
count   8469.000000                   2769.000000
mean      44.026804                      2.991333
std       15.296112                      1.407016
min       18.000000                      1.000000
25%       31.000000                      2.000000
50%       44.000000                      3.000000
75%       57.000000                      4.000000
max       70.000000                      5.000000

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8469 entries, 1 to 8469
Data columns (total 14 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Customer Age                  8469 non-null   int64  
 1   Customer Gender               8469 non-null   object 
 2   Product Purchased             8469 non-null   object 
 3   Date of Purchase              8469 non-null   object 
 4   Ticket Type                   8469 non-null   object 
 5   Ticket Subject                8469 non-null   object 
 6   Ticket Description            8469 non-null   object 
 7   Ticket Status                 8469 non-null   object 
 8   Resolution                    2769 non-null   object 
 9   Ticket Priority               8469 non-null   object 
 10  Ticket Channel                8469 non-null   object 
 11  First Response Time           5650 non-null   object 
 12  Time to Resolution            2769 non-null   object 
 13  Customer

Looking at the 'Resolution' column, we can see that the responses are largely gibberish, so this column serves little useful purpose and should also be deleted.

In [9]:
df.drop('Resolution', axis=1, inplace=True)
df.head()

Customer Age Customer Gender Product Purchased Date of Purchase  \
Ticket ID                                                                    
1                    32           Other        GoPro Hero       2021-03-22   
2                    42          Female       LG Smart TV       2021-05-22   
3                    48           Other          Dell XPS       2020-07-14   
4                    27          Female  Microsoft Office       2020-11-13   
5                    67          Female  Autodesk AutoCAD       2020-02-04   

               Ticket Type            Ticket Subject  \
Ticket ID                                              
1          Technical issue             Product setup   
2          Technical issue  Peripheral compatibility   
3          Technical issue           Network problem   
4          Billing inquiry            Account access   
5          Billing inquiry                 Data loss   

                                          Ticket Description  \
Ticket ID                                                      
1          I'm having an issue with the {product_purchase...   
2          I'm having an issue with the {product_purchase...   
3          I'm facing a problem with my {product_purchase...   
4          I'm having an issue with the {product_purchase...   
5          I'm having an issue with the {product_purchase...   

                       Ticket Status Ticket Priority Ticket Channel  \
Ticket ID                                                             
1          Pending Customer Response        Critical   Social media   
2          Pending Customer Response        Critical           Chat   
3                             Closed             Low   Social media   
4                             Closed             Low   Social media   
5                             Closed             Low          Email   

           First Response Time   Time to Resolution  \
Ticket ID                                             
1          2023-06-01 12:15:36                  NaN   
2          2023-06-01 16:45:38                  NaN   
3          2023-06-01 11:14:38  2023-06-01 18:05:38   
4          2023-06-01 07:29:40  2023-06-01 01:57:40   
5          2023-06-01 00:12:42  2023-06-01 19:53:42   

           Customer Satisfaction Rating  
Ticket ID                                
1                                   NaN  
2                                   NaN  
3                                   3.0  
4                                   3.0  
5                                   1.0

Some initial explorations I want to consider:

What is the most common ticket type?
What is the most common ticket channel?
Which products have the highest number of technical issues?
Is there a correlation between the age / gender of the customer and the ticket channel they use? 
Is there a correlation between the ticket subject, the ticket priority and the ticket channel used?
Is there a correlation between response time and customer satisfaction rating?
Do the ticket subject, ticket priority and ticket channel have an impact on response time?

Looking over these questions and the data again, and it looks like Ticket Description will not be a particulalry useful column for analysis. 

I have also not yet made any questions relating to Date of Purchase. It may be useful to know when the product was purchased to get an idea of how long it takes for users to have issues, so we will keep this column for now, but it may be removed later.

In [10]:
df.drop('Ticket Description', axis=1, inplace=True)
df.head()

Customer Age Customer Gender Product Purchased Date of Purchase  \
Ticket ID                                                                    
1                    32           Other        GoPro Hero       2021-03-22   
2                    42          Female       LG Smart TV       2021-05-22   
3                    48           Other          Dell XPS       2020-07-14   
4                    27          Female  Microsoft Office       2020-11-13   
5                    67          Female  Autodesk AutoCAD       2020-02-04   

               Ticket Type            Ticket Subject  \
Ticket ID                                              
1          Technical issue             Product setup   
2          Technical issue  Peripheral compatibility   
3          Technical issue           Network problem   
4          Billing inquiry            Account access   
5          Billing inquiry                 Data loss   

                       Ticket Status Ticket Priority Ticket Channel  \
Ticket ID                                                             
1          Pending Customer Response        Critical   Social media   
2          Pending Customer Response        Critical           Chat   
3                             Closed             Low   Social media   
4                             Closed             Low   Social media   
5                             Closed             Low          Email   

           First Response Time   Time to Resolution  \
Ticket ID                                             
1          2023-06-01 12:15:36                  NaN   
2          2023-06-01 16:45:38                  NaN   
3          2023-06-01 11:14:38  2023-06-01 18:05:38   
4          2023-06-01 07:29:40  2023-06-01 01:57:40   
5          2023-06-01 00:12:42  2023-06-01 19:53:42   

           Customer Satisfaction Rating  
Ticket ID                                
1                                   NaN  
2                                   NaN  
3                                   3.0  
4                                   3.0  
5                                   1.0

Some other observations at this stage:

1) Not specified in the original data, but we might assume that the data was collected via a link sent after the ticket was resolved. Else it seems strange to be collecting info on age and gender just from social media / email, for example.

2) We should establish how many responses are from Male, Female, Other. 'Other' might refer to nonbinary/GNC responders, but improbable that all 'Other' responses classify as this. May just be due to how customers filled out the response forms.

3) Time to Resolution is likely NaN for any incidents that are still open. Is the same true for Customer Satisfaction Rating? Should check.

4) First Response Time and Time to Resolution are not particularly helpful on their own. Data should have also been collected for when the customer first logged the incident, to establish the wait time and help with Service Level Agreeement (SLA) monitoring. Time to Resolution seems more like 'Time OF Resolution', we could add an additional column to track the time to resolution and see how it lines up with Satisfaction Rating etc.

In [11]:
df['Ticket Status'].nunique()

3

In [12]:
print(df['Ticket Status'].unique())

['Pending Customer Response' 'Closed' 'Open']


In [13]:
print(df['Ticket Status'].value_counts())

Ticket Status
Pending Customer Response    2881
Open                         2819
Closed                       2769
Name: count, dtype: int64


This shows us how many tickets are in each of the three statuses: Pending Customer Response, Open, and Closed. I expect that 'Open' tickets are likely waiting for action from the helpdesk, else they would be in 'Pending Customer Response'. I could find out the exact percentage of each ticket status, but from a glance we can see that they are all roughly equal, so a third each.

5,700 total open tickets seems like quite a high number compared to 2769 closed tickets. It may be worth examining the most common types/subjects of open ticket, as well as the channel used, to begin to examine how we might reduce the number of open tickets on the system.

In [14]:
most_common_product = df['Product Purchased'].value_counts().idxmax()
print("The most common product is:", most_common_product)

The most common product is: Canon EOS


In [15]:
print(df['Product Purchased'].value_counts())

Product Purchased
Canon EOS                         240
GoPro Hero                        228
Nest Thermostat                   225
Philips Hue Lights                221
Amazon Echo                       221
LG Smart TV                       219
Sony Xperia                       217
Roomba Robot Vacuum               216
Apple AirPods                     213
LG OLED                           213
iPhone                            212
Sony 4K HDR TV                    210
Garmin Forerunner                 208
LG Washing Machine                208
Canon DSLR Camera                 206
Nikon D                           204
Nintendo Switch Pro Controller    203
Google Pixel                      203
Fitbit Charge                     202
Sony PlayStation                  202
HP Pavilion                       200
Microsoft Office                  200
Amazon Kindle                     198
Google Nest                       198
Dyson Vacuum Cleaner              198
Bose SoundLink Speaker          

It may be worth adding an additional column that has a Category for each type of product.'Camera', 'Smartphone', 'Gaming' etc. 

It's interesting that there is not a great deal of variance between the highest and lowest number of tickets raised. This makes me wonder if the data collected was only for the products with the most tickets raised, rather than all products.

Let's do some value counts for some of our other columns to get a better understanding of the data.

In [16]:
print(df['Customer Gender'].value_counts())

Customer Gender
Male      2896
Female    2887
Other     2686
Name: count, dtype: int64


In [17]:
print(df['Ticket Type'].value_counts())

Ticket Type
Refund request          1752
Technical issue         1747
Cancellation request    1695
Product inquiry         1641
Billing inquiry         1634
Name: count, dtype: int64


In [18]:
print(df['Ticket Subject'].value_counts())

Ticket Subject
Refund request              576
Software bug                574
Product compatibility       567
Delivery problem            561
Hardware issue              547
Battery life                542
Network problem             539
Installation support        530
Product setup               529
Payment issue               526
Product recommendation      517
Account access              509
Peripheral compatibility    496
Data loss                   491
Cancellation request        487
Display issue               478
Name: count, dtype: int64


In [19]:
print(df['Ticket Priority'].value_counts())

Ticket Priority
Medium      2192
Critical    2129
High        2085
Low         2063
Name: count, dtype: int64


In [20]:
print(df['Ticket Channel'].value_counts())

Ticket Channel
Email           2143
Phone           2132
Social media    2121
Chat            2073
Name: count, dtype: int64


As we can see, each of the columns has low variance between its highest and lowest value. My instinct is to question whether or not this data is actually real, as mentioned previously. I'll continue to work with this dataset, but it may be worth bearing in mind that it might not be representative of a real environment.

In [21]:
top_products = ['Canon EOS', 'GoPro Hero', 'Nest Thermostat', 'Philips Hue Lights', 'Amazon Echo']

top_5 = df[df['Product Purchased'].isin(top_products)]

product_ticket_counts = top_5.groupby(['Product Purchased', 'Ticket Status']).size().unstack(fill_value=0)

print(product_ticket_counts)


Ticket Status       Closed  Open  Pending Customer Response
Product Purchased                                          
Amazon Echo             72    68                         81
Canon EOS               83    90                         67
GoPro Hero              80    72                         76
Nest Thermostat         73    70                         82
Philips Hue Lights      68    87                         66


Of the top 5 products, it looks like the most common, the Canon EOS, also has the most closed tickets, but also the most open tickets. It has the second lowest number of 'Pending Customer response' out of these top 5 products, however. 

In [22]:
canon_eos_entries = df[df['Product Purchased'] == 'Canon EOS']
print(canon_eos_entries[['Ticket Type', 'Ticket Subject']])

                    Ticket Type            Ticket Subject
Ticket ID                                                
20               Refund request              Software bug
37               Refund request            Hardware issue
86              Product inquiry  Peripheral compatibility
145              Refund request                 Data loss
200             Product inquiry              Software bug
...                         ...                       ...
8223            Billing inquiry              Software bug
8230       Cancellation request              Battery life
8325             Refund request     Product compatibility
8342            Billing inquiry          Delivery problem
8350            Billing inquiry            Account access

[240 rows x 2 columns]


In [23]:
print(canon_eos_entries['Ticket Type'].value_counts())

Ticket Type
Product inquiry         52
Refund request          50
Technical issue         48
Billing inquiry         46
Cancellation request    44
Name: count, dtype: int64


In [24]:
print(canon_eos_entries['Ticket Subject'].value_counts())

Ticket Subject
Battery life                22
Hardware issue              20
Account access              19
Delivery problem            18
Product compatibility       18
Display issue               17
Software bug                16
Peripheral compatibility    16
Product recommendation      15
Installation support        14
Payment issue               14
Product setup               12
Data loss                   11
Refund request              10
Network problem             10
Cancellation request         8
Name: count, dtype: int64


For the Canon EOS, it looks like battery life issues or other issues relating to hardware are the most common issue. It's difficult to establish what 'account access' might mean - maybe we'll delve deeper into this shortly.

We can see some issues when looking at the canon_eos_entries Ticket Type and Ticket Subject. "Billing inquiry - Software bug" isn't the most logical of combinations, for example. There could be two reasons why we are seeing results like this:

1) The customer is asked to select their Ticket Type and Ticket Subject. If this is the case, then my feedback would be that the Customer UX needs to be updated, for example so that if they select 'Billing inquiry' then 'Software bug' is not given as a vialble Subject. This would make it easier to report on the data, but also for the customer to get the relevant help for their product.

2) The ticket type and ticket subject are completed by the helpdesk agent. If this is true, then agents need to be more careful with what Subject they pair with each Type. Again, this can impact reporting and make it harder for customers to receive the correct help.

At this stage, I want to work out the range of dates this data was sourced from. 

In [25]:
df['First Response Time'] = pd.to_datetime(df['First Response Time'])
earliest_date = df['First Response Time'].min()
latest_date = df['First Response Time'].max()

print("Earliest Date:", earliest_date)
print("Latest Date:", latest_date)

Earliest Date: 2023-05-31 21:55:39
Latest Date: 2023-06-02 00:54:21


I converted the values in the 'First Response Time' to dates and times - we'll do the ones for Last Response Time later. 

We can see that all of the tickets in this dataset were raised over the course of a small space of time - between 21:55 on the 31st May 2023, and 00:54 on the 2nd June. 

If we wanted to get more granular, we could work out the number of tickets logged each hour during this time period, and whether they were via Chat, Email, Phone or Social Media.

In [26]:
df.head()

Customer Age Customer Gender Product Purchased Date of Purchase  \
Ticket ID                                                                    
1                    32           Other        GoPro Hero       2021-03-22   
2                    42          Female       LG Smart TV       2021-05-22   
3                    48           Other          Dell XPS       2020-07-14   
4                    27          Female  Microsoft Office       2020-11-13   
5                    67          Female  Autodesk AutoCAD       2020-02-04   

               Ticket Type            Ticket Subject  \
Ticket ID                                              
1          Technical issue             Product setup   
2          Technical issue  Peripheral compatibility   
3          Technical issue           Network problem   
4          Billing inquiry            Account access   
5          Billing inquiry                 Data loss   

                       Ticket Status Ticket Priority Ticket Channel  \
Ticket ID                                                             
1          Pending Customer Response        Critical   Social media   
2          Pending Customer Response        Critical           Chat   
3                             Closed             Low   Social media   
4                             Closed             Low   Social media   
5                             Closed             Low          Email   

          First Response Time   Time to Resolution  \
Ticket ID                                            
1         2023-06-01 12:15:36                  NaN   
2         2023-06-01 16:45:38                  NaN   
3         2023-06-01 11:14:38  2023-06-01 18:05:38   
4         2023-06-01 07:29:40  2023-06-01 01:57:40   
5         2023-06-01 00:12:42  2023-06-01 19:53:42   

           Customer Satisfaction Rating  
Ticket ID                                
1                                   NaN  
2                                   NaN  
3                                   3.0  
4                                   3.0  
5                                   1.0

In [27]:
df['Time to Resolution'] = pd.to_datetime(df['Time to Resolution'])
earliest_date = df['Time to Resolution'].min()
latest_date = df['Time to Resolution'].max()

print("Earliest Date:", earliest_date)
print("Latest Date:", latest_date)

Earliest Date: 2023-05-31 21:53:30
Latest Date: 2023-06-02 00:55:33


In [28]:
df.rename(columns={'Time to Resolution': 'Time of Resolution'})
df.head()

Customer Age Customer Gender Product Purchased Date of Purchase  \
Ticket ID                                                                    
1                    32           Other        GoPro Hero       2021-03-22   
2                    42          Female       LG Smart TV       2021-05-22   
3                    48           Other          Dell XPS       2020-07-14   
4                    27          Female  Microsoft Office       2020-11-13   
5                    67          Female  Autodesk AutoCAD       2020-02-04   

               Ticket Type            Ticket Subject  \
Ticket ID                                              
1          Technical issue             Product setup   
2          Technical issue  Peripheral compatibility   
3          Technical issue           Network problem   
4          Billing inquiry            Account access   
5          Billing inquiry                 Data loss   

                       Ticket Status Ticket Priority Ticket Channel  \
Ticket ID                                                             
1          Pending Customer Response        Critical   Social media   
2          Pending Customer Response        Critical           Chat   
3                             Closed             Low   Social media   
4                             Closed             Low   Social media   
5                             Closed             Low          Email   

          First Response Time  Time to Resolution  \
Ticket ID                                           
1         2023-06-01 12:15:36                 NaT   
2         2023-06-01 16:45:38                 NaT   
3         2023-06-01 11:14:38 2023-06-01 18:05:38   
4         2023-06-01 07:29:40 2023-06-01 01:57:40   
5         2023-06-01 00:12:42 2023-06-01 19:53:42   

           Customer Satisfaction Rating  
Ticket ID                                
1                                   NaN  
2                                   NaN  
3                                   3.0  
4                                   3.0  
5                                   1.0

In [29]:
morning_mask = (df['First Response Time'].dt.hour >= 7) & (df['First Response Time'].dt.hour < 12)
morning = df[morning_mask]
morning_entries = morning.shape[0]
print("There were {} entries raised between 7AM and 12PM in this dataset.".format(morning_entries))

There were 1167 entries raised between 7AM and 12PM in this dataset.


In [30]:
print(morning['Ticket Channel'].value_counts())

Ticket Channel
Chat            308
Phone           302
Email           285
Social media    272
Name: count, dtype: int64


In [31]:
noon_mask = (df['First Response Time'].dt.hour >= 12) & (df['First Response Time'].dt.hour < 17)
noon = df[noon_mask]
noon_entries = noon.shape[0]
print("There were {} entries raised between 12PM and 5PM in this dataset.".format(noon_entries))

There were 1123 entries raised between 12PM and 5PM in this dataset.


In [32]:
evening_mask = (df['First Response Time'].dt.hour >= 17) & (df['First Response Time'].dt.hour < 21)
evening = df[evening_mask]
evening_entries = evening.shape[0]
print("There were {} entries raised between 5PM and 9PM in this dataset.".format(evening_entries))

There were 982 entries raised between 5PM and 9PM in this dataset.


In [33]:
night_mask = ((df['First Response Time'].dt.hour >= 21) & (df['First Response Time'].dt.hour <= 23)) | (df['First Response Time'].dt.hour < 7)
night= df[night_mask]
night_entries = df[night_mask].shape[0]
print("There were {} entries raised between 9PM and 7AM in this dataset.".format(night_entries))

#Python didn't want to return results when I only used one condition, so I used two: one to show between 9PM and Midnight, and another for Midnight to 7 AM. Then used the OR condition to return both.

There were 2378 entries raised between 9PM and 7AM in this dataset.


In [34]:
all_entries_mask = ~df.index.isin(morning.index) & ~df.index.isin(noon.index) & ~df.index.isin(evening.index) & ~df.index.isin(night.index)
remaining_entries = df[all_entries_mask]

print("Entries not in 'morning', 'noon', 'evening', or 'night' dataframes:")
print(remaining_entries)

#This code will tell me why I have so many missing entries, by working out what entries from df are NOT in any of the time-based frames I created.

Entries not in 'morning', 'noon', 'evening', or 'night' dataframes:
           Customer Age Customer Gender       Product Purchased  \
Ticket ID                                                         
6                    53            Male        Microsoft Office   
7                    24           Other       Microsoft Surface   
8                    23            Male      Philips Hue Lights   
19                   19            Male      LG Washing Machine   
23                   54          Female                    Xbox   
...                 ...             ...                     ...   
8462                 28          Female        Microsoft Office   
8464                 54           Other      LG Washing Machine   
8465                 22          Female                 LG OLED   
8466                 27          Female  Bose SoundLink Speaker   
8469                 53           Other      Philips Hue Lights   

          Date of Purchase           Ticket Type            

In [35]:
remaining_entries_with_nat = remaining_entries[remaining_entries['First Response Time'].isna()]
num_entries_with_nat = remaining_entries_with_nat.shape[0]

print("Number of entries with NaT in 'First Response Time':", num_entries_with_nat)

Number of entries with NaT in 'First Response Time': 2819


In [36]:
na_response_not_open_mask = df['First Response Time'].isna() & (df['Ticket Status'] != 'Open')
num_entries_na_response_not_open = df[na_response_not_open_mask].shape[0]

print("Number of entries where 'First Response Time' is NaT and 'Ticket Status' is not 'Open':", num_entries_na_response_not_open)

Number of entries where 'First Response Time' is NaT and 'Ticket Status' is not 'Open': 0


From this experiment, we can see that it looks like 'First Response Time' is likely the time the helpdesk responds to the incident, and not when the incident was originally logged.

If this was a real dataset, this would be concerning. We need to know when customers are logging incidents so that we can plan appropriately. Not having that data could have a critical impact on customer service and business operation. What's the point of having lots of agents working nights if no customers are raising incidents then? 

I'd also be concerned about the number of contacts being made to customers between the hours of 9PM and 7AM, but as we've already established, the actual data provided is likely just made up. I'd hope that no service desk would make such an error... 

It's also worth remembering that the data is taken from a short timespan of roughly three days, so it may not be representative of the average service desk functionality.

In [37]:
channel_counts = df.groupby(['Customer Gender', 'Ticket Channel']).size().reset_index(name='Count')
max_counts = channel_counts.groupby('Customer Gender')['Count'].idxmax()
channels_per_gender = channel_counts.loc[max_counts, ['Customer Gender', 'Ticket Channel', 'Count']]
channels_per_gender

Customer Gender Ticket Channel  Count
1          Female          Email    749
7            Male   Social media    737
9           Other          Email    688

In [38]:
average_age_by_channel = df.groupby('Ticket Channel')['Customer Age'].mean().reset_index(name='Average Age')
average_age_by_channel['Average Age'] = average_age_by_channel['Average Age'].round(0).astype(int)
average_age_by_channel

Ticket Channel  Average Age
0           Chat           44
1          Email           44
2          Phone           44
3   Social media           43

No real surprises here, the data is very evenly spread. 

Next steps: work out how to show the actual resolution time and perform analysis.